### DISCUSSION TITLE
how to construct product name from reactant names in EnumerateLibrary()?

### DISCUSSION THREAD

**Discussion by rczerminski-valo** [rdkit/rdkit#4771] [General]
I just started using ```Chem.rdChemReactions.EnumerateLibrary(rxn, [a, b])``` and it looks great! Is there some way to instruct  `EnumerateLibrary` to carry names from building blocks to the products? As an example, assume I have two component reaction as above where `a` and `b` are lists of molecules (e.g. a = [a1, a2], b = [b1,b2]) and each molecule has "name" property set (e.g. a1.GetProp("name") = "a1", etc) in the final products (pij = ai + bj) I would like to be able to have: pij.GetProp("name") = "ai.bj", etc
I can do it using `RunReactants()`, looping of over reactants, etc. but it would be nice to have as a built-in `EnumerateLibrary()` functionality.

**Comment by bp-kelley** [comment]
It is possible to get the index of the reagents used for making the products using EnumerateLibrary.GetPosition

A roll your own solution might be as follows:

```
bbs = ...
en = EnumerateLibrary(...)
for prods in en:
     position = en.GetPosition()
     names = [bbs[reactant_idx][bb_idx].GetProp("name") for reactant_idx, bb_idx in enumerate(position)]
     name = ".".join(names)
     for p in prods:
          p.SetProp("name", name)
```

A fully working (and contrived) example is here:

```
from rdkit import Chem
from rdkit.Chem import rdChemReactions

smirks_thiourea = "[N;$(N-[#6]):3]=[C;$(C=S):1].[N;$(N[#6]);!$(N=*);!$([N-]);!$(N#*);!$([ND3]);!$([ND4]);!$(N[O,N]);!$(N[C,S]=[S,O,N]):2]>>[N:3]-[C:1]-[N+0:2]"
rxn = rdChemReactions.ReactionFromSmarts(smirks_thiourea)
reagents = [
    [Chem.MolFromSmiles('C=CCN=C=S'), Chem.MolFromSmiles('CC=CCN=C=S')],
    [Chem.MolFromSmiles('NCc1ncc(Cl)cc1Br'),
     Chem.MolFromSmiles('NCCc1ncc(Cl)cc1Br'),
     Chem.MolFromSmiles('NCCCc1ncc(Cl)cc1Br'),
    ]
]

# set the names of the reagents
for i, bbs in enumerate(reagents):
    for j, bb in enumerate(bbs):
        bb.SetProp("name", f"reagent{i}:num{j}")

enumerator = rdChemReactions.EnumerateLibrary(rxn, reagents)

for prods in enumerator:
    # get the reagents used and concatenate their names
    pos = enumerator.GetPosition()
    names = [reagents[reactant_idx][bb_idx].GetProp("name") for reactant_idx, bb_idx in enumerate(pos)]
    name = ".".join(names)
    for mols in prods:
        for mol in mols:
            mol.SetProp("name", name)
            print(Chem.MolToSmiles(mol), mol.GetProp("name"))
```

the output is

```
C=CCNC(=S)NCc1ncc(Cl)cc1Br reagent0:num0.reagent1:num0
CC=CCNC(=S)NCc1ncc(Cl)cc1Br reagent0:num1.reagent1:num0
C=CCNC(=S)NCCc1ncc(Cl)cc1Br reagent0:num0.reagent1:num1
CC=CCNC(=S)NCCc1ncc(Cl)cc1Br reagent0:num1.reagent1:num1
C=CCNC(=S)NCCCc1ncc(Cl)cc1Br reagent0:num0.reagent1:num2
CC=CCNC(=S)NCCCc1ncc(Cl)cc1Br reagent0:num1.reagent1:num2
```

**Comment by bp-kelley** [comment]
Having this functionality in the enumerator itself is a good idea, I'll think about how to enable it.

**Comment by palazzef** [comment]
Hi,
Sorry if I re-open this old thread, but I have tried the proposed suggestion and it didn't work: the outputting reagent is not the correct one. Indeed it's the second reagent (the one with Br) that reacts with "mol" (that instead is correctly outputted). Could you please help me in understanding where is the error? I am using rdkit 2024.03.3.
Thanks a lot!

from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import rdChemReactions

reaction = 'O-[c;H0;D3;+0:1](:[c:2]):[c:3].[O;D1;H0:4]=[C:5]-[OH;D1;+0:6]>>[O;D1;H0:4]=[C:5]-[O;H0;D2;+0:6]-[c;H0;D3;+0:1](:[c:2]):[c:3]'
reagents = [Chem.MolFromSmiles('ClC(=O)c1ccc(B=O)cc1'), Chem.MolFromSmiles('OC(=O)C(Br)=C(Br)C(O)=O'), Chem.MolFromSmiles('ClC(=O)c1ccc(B=O)cc1')]
mol = Chem.MolFromSmiles('CC(CC1=CC2=C(C=C1)OCO2)NC')

rxn = rdChemReactions.ReactionFromSmarts(reaction)
rxn.Initialize()
reagents = [[mol], reagents]
for i, bbs in enumerate(reagents):
    for j, bb in enumerate(bbs):
        bb.SetProp("name", f"{Chem.MolToSmiles(bb)}")

enumerator  = rdChemReactions.EnumerateLibrary(rxn, reagents)

for prods in enumerator:
    pos = enumerator.GetPosition()
    names = [reagents[reactant_idx][bb_idx].GetProp("name") for reactant_idx, bb_idx in enumerate(pos)]
    name = ".".join(names)
    for mols in prods:
     for mol in mols:
      print(Chem.MolToSmiles(mol), name)

Outcome:
CNC(C)Cc1ccc(OC(=O)C(Br)=C(Br)C(=O)O)c(OC)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC(=O)C(Br)=C(Br)C(=O)O)c(OC)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC(=O)C(Br)=C(Br)C(=O)O)c(OC)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC(=O)C(Br)=C(Br)C(=O)O)c(OC)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC)c(OC(=O)C(Br)=C(Br)C(=O)O)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC)c(OC(=O)C(Br)=C(Br)C(=O)O)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC)c(OC(=O)C(Br)=C(Br)C(=O)O)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
CNC(C)Cc1ccc(OC)c(OC(=O)C(Br)=C(Br)C(=O)O)c1 CNC(C)Cc1ccc2c(c1)OCO2.O=Bc1ccc(C(=O)Cl)cc1
